Implement searchScopes

AaronFriel opened the pull request:
<!--
Thank you for submitting a pull request!

Here's a checklist you might find useful.
[ ] There is an associated issue that is labeled
  'Bug' or 'help wanted' or is in the Community milestone
[X] Code is up-to-date with the `master` branch
[X] You've successfully run `jake runtests` locally
[X] You've signed the CLA
[X] There are new or updated unit tests validating the change

Refer to CONTRIBUTING.MD for more details.
  https://github.com/Microsoft/TypeScript/blob/master/CONTRIBUTING.md
-->

Fixes #23952

Implements a new compiler option called `searchScopes`.

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/Microsoft/TypeScript?pullRequest=24075) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/Microsoft/TypeScript?pullRequest=24075) before we can accept your contribution.<br/><br/>:x: AaronFriel [sign now](https://cla.opensource.microsoft.com/Microsoft/TypeScript?pullRequest=24075)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/Microsoft/TypeScript?pullRequest=24075) it.</sub>

AaronFriel commented:
I was pleased at how easily I could navigate the TypeScript code base! I earned back the code in the first commit by verifying (via `git checkout HEAD~2 ...`) that the tests check for the new behavior.

Replying to #23952 commenters @weswigham and @mhegazy, and CCing @DanielRosenwasser whom I spoke to at Build about this. I believe I've implemented the feature I'm requesting here. Or at least I've implemented the `tsconfig` portion of this, right?

I think I still need to implement something in `vscode` for the automatic type acquisition, and not just the use of additional search scopes?

Of course, this feature may not end up being accepted. I would be disappointed by that but I do hope an alternative would be proposed to support alternative scope searching for private NPM repositories and to support alternatives to Definitely Typed.

mhegazy commented on src/compiler/types.ts:
```diff
@@ -4331,6 +4331,8 @@ namespace ts {
         /** Paths used to compute primary types search locations */
         typeRoots?: string[];
+        /** Scopes used to search for type-only packages */
+        searchScopes?: string[];
```
I think we need to talk about where this option lives, in the tsconfig or a sever commandline option that users specify in IDE configuration for instance..

if it is in tsconfig.json then it needs to be added to the commandline-parser to validate it..

if it a server option, then it needs to be in the server.ts instead.

AaronFriel replied:
Agreed, but if it is not in the `tsconfig.json` then how will the compiler when invoked from the command line pull in typings from `node_modules/@companyTypes/companyPackageName`?

mhegazy replied:
is this for the commandline for for ATA (i.e. editor only scenarios)? may be you can explain the request one more time in more details.

AaronFriel replied:
Both. I'd like to implement both pieces.

mhegazy commented on src/compiler/moduleNameResolver.ts:
```diff
@@ -150,31 +150,41 @@ namespace ts {
 
     /**
-     * Returns the path to every node_modules/@types directory from some ancestor directory.
+     * Returns the path to every type search scope directory from some ancestor directory.
```
i am actually not sure what the intent of this feature is now.. [I thought](https://github.com/Microsoft/TypeScript/issues/23952#issuecomment-387573894) this was manly for .js file intellisense in VSCode, and you wanted the server to download types from `@Company\types` instead of `@types`.. 

if that is the case, then no changes are needed here, changes should be done to the typinging installer, and to the tsserver module resolver.

AaronFriel replied:
If this is not changed, then how will types installed via `npm install --save-dev @company-types/companyPackageName` work in the same manner as `npm install --save-dev @types/packageName`?

mhegazy replied:
so why not use typeRoots then?

AaronFriel replied:
Because that would break the default type acquisition. I want to *supplement* it, not rip it to shreds! From the docs, emphasis mine:

By default all *visible* "`@types`" packages are included in your compilation.
Packages in `node_modules/@types` of any enclosing folder are considered *visible*;
specifically, that means packages within `./node_modules/@types/`,  `../node_modules/@types/`, `../../node_modules/@types/`, and so on.

***If `typeRoots` is specified, *only* packages under `typeRoots` will be included.
For example:***

```json
{
   "compilerOptions": {
       "typeRoots" : ["./typings"]
   }
}
```

This config file will include *all* packages under `./typings`, and no packages from `./node_modules/@types`.

AaronFriel replied:
I repeat:

# If typeRoots is specified, only packages under typeRoots will be included.
For example:

# This config file will include all packages under ./typings, and no packages from ./node_modules/@types.

AaronFriel replied:
The `if` block on lines 140-142 are why I am implementing a new option and not using `typeRoots`. If I set `typeRoots`, I opt out of the standard behavior for finding typings and completions in the editor experience.

https://github.com/Microsoft/TypeScript/blob/77168e5d500269ab6c044d688b53255077fa9201/src/compiler/moduleNameResolver.ts#L139-L155

AaronFriel replied:
**I do not want to opt out of the standard behavior of finding typings, I want to add to it and allow the users to specify additional scopes that are treated just like `@types`.**

AaronFriel replied:
I am sorry I am so frustrated but I feel like I have been as clear as I can be about the use case, setting `typeRoots` short-circuits `getEffectiveTypeRoots`, which changes the behavior of the compiler and editor functionality. It changes the behavior in the module name resolver, resolution cache, language server module name finding and code completions. Search for every place `getEffectiveTypeRoots` is called. All of those call sites are short-circuited to return the fixed list the user provided instead of the recursive searching done by default.

weswigham replied:
> fixed list the user provided instead of the recursive searching done by default.

Ah. So you're saying you rely on the directory traversal to find `@types` (or corporate `@companyname-types`) folders rather than relying, eg, on a single top-level `node_modules/@types` and `node_modules/@companyname-types` which you could easily include in your `typeRoots` list. Got it.

AaronFriel replied:
YES! Thank you! I am so sorry it took so many words to express this.

weswigham replied:
No worries; we were wondering about the utility of the same when we initially implemented `typeRoots` a few years ago.

mhegazy replied:
So how many node_modules do you have in your project typically? I ask cause if they are one or two, typeRoots should do the trick.

Also sorry if I am asking questions, but u originally said ATA and that is a totally different feature.

AaronFriel replied:
I want it to work just like `@types` does, full stop. I suspect this will take two sets of changes, the first here to make the compiler treat `searchScopes` as an alternate source for the string "@types", and I'm not quite sure about how to implement ATA.

I think for ATA to work correctly, I will need to modify the `npm` command the server runs?

mhegazy replied:
Can you tell us more about your use case? how does this come into play in your organization? and why typeRoots is not sufficient for your use case?

AaronFriel replied:
The use case is an internal NPM-compatible package registry, and wanting to put typings in it to avoid placing them in upstream projects, and still support automatic type acquisition.

I don't like typeroots because (a.) it changes the default type resolution algorithm `getEffectiveTypeRoots`, and I don't know what packages might depend on that and there must be *some* reason for walking the spine of the directory in the default behavior, and I don't want to change that behavior, and (b.) it doesn't enable automatic type acquisition.

AaronFriel replied:
To be clear:

1. Setting `typeRoots: ["node_modules/@types"]` is not a no-op. I don't know why it's not a no-op, I couldn't find a rationale for the `@types` scope behavior, but it *is*. That's just how it is.

2. Setting `typeRoots: ["node_modules/@types"]` could subtly break compilation because of that. I don't know under what situations that would happen - I suspect in monorepo (`lerna` or `yarn` workspaces) projects it could cause `./node_modules/@types` installed at the root not to resolve in child projects. Whatever the case, the fact is that setting this changes behavior and could cause projects to not compile for failure to resolve @types. If that isn't true, then there's no reason for `getEffectiveTypeRoots` to short circuit.

3. Setting `typeRoots: ["node_modules/@companyTypes", "node_modules/@types"]` does not enable automatic type acquisition for the scope `@companyTypes`, and could subtly break compilation as above.

lenovouser commented:
I'd need this too. I have exactly the same setup. A private registry with types that I'd like to supplement the `@types` with. Also this is very nice for testing types before pushing them to `definitelytyped/definitelytyped` too
